zircon.core/common/src/main/kotlin/org/hexworks/zircon/internal/component/impl/DefaultNumberInput.kt

```diff
@@ -0,0 +1,259 @@
+            render()
+            Processed
+        } else Pass
+    }
+
+    override fun mousePressed(event: MouseEvent, phase: UIEventPhase) = whenEnabledRespondWith {
+        if (phase == UIEventPhase.TARGET) {
+            componentStyleSet.applyActiveStyle()
+            render()
+            Processed
+        } else Pass
+    }
+
+    override fun mouseReleased(event: MouseEvent, phase: UIEventPhase) = whenEnabledRespondWith {
+        if (phase == UIEventPhase.TARGET) {
+            componentStyleSet.applyMouseOverStyle()
+            render()
+            Processed
+        } else Pass
+    }
+
+    override fun keyPressed(event: KeyboardEvent, phase: UIEventPhase) = whenEnabledRespondWith {
+        if (phase == UIEventPhase.TARGET) {
+            if (isNavigationKey(event)) {
+                Pass
+            } else {
+                when (event.code) {
+                    KeyCode.ENTER -> {
+                        saveModifications()
+                        clearFocus()
+                    }
+                    KeyCode.ESCAPE -> clearFocus()
+                    KeyCode.RIGHT -> textBuffer.applyTransformation(MoveCursor(RIGHT))
+                    KeyCode.LEFT -> textBuffer.applyTransformation(MoveCursor(LEFT))
+                    KeyCode.DELETE -> textBuffer.applyTransformation(DeleteCharacter(DEL))
+                    KeyCode.BACKSPACE -> textBuffer.applyTransformation(DeleteCharacter(BACKSPACE))
+                    else -> {
+                        event.key.forEach { char ->
+                            if (TextUtils.isDigitCharacter(char)) {
+                                checkAndAddChar(char)
+                            }
+                        }
+                    }
+                }
+                refreshCursor()
+                render()
+                Processed
+            }
+        } else Pass
+    }
+
+    override fun render() {
+        renderingStrategy.render(this, graphics)
+    }
+
+    private fun isNavigationKey(event: KeyboardEvent) =
+            event == TAB || event == REVERSE_TAB
+
+    private fun checkAndAddChar(char: Char) {
+        val virtualTextBuffer = EditableTextBuffer.create(text, textBuffer.cursor)
+        virtualTextBuffer.applyTransformation(InsertCharacter(char))
+        if (virtualTextBuffer.getText().toInt() <= maxValue) {
+            if (text.length == maxNumberLength) {
+                textBuffer.applyTransformation(DeleteCharacter(BACKSPACE))
+            }
+            textBuffer.applyTransformation(InsertCharacter(char))
+        } else {
+            if (textBuffer.getCharAt(textBuffer.cursor.position).isPresent) {
+                textBuffer.applyTransformation(DeleteCharacter(DEL))
+            } else {
+                textBuffer.applyTransformation(DeleteCharacter(BACKSPACE))
+            }
+            checkAndAddChar(char)
+        }
+    }
+
+    private fun computeDigitalValue() {
+        currentValue = if (text == "") {
```

Comment:
I like that you used the fact that `if` is an expression and not a statement in Kotlin! 👍